app/src/heartbeat/heartbeat_controller_desktop.cc

```diff
@@ -79,5 +97,171 @@ void HeartbeatController::LogHeartbeat() {
   scheduler_.Schedule(log_heartbeat_funct);
 }
 
+std::string HeartbeatController::GetAndResetStoredHeartbeats() {
+  std::shared_ptr<std::string> output_str = std::make_shared<std::string>("");
+  std::shared_ptr<Semaphore> scheduled_work_semaphore =
+      std::make_shared<Semaphore>(0);
+
+  std::function<void(void)> get_and_reset_function = [&, output_str]() {
+    std::string current_date = date_provider_.GetDate();
+    // Return early if a heartbeat has already been logged today.
+    if (this->last_fetched_all_heartbeats_date_ != current_date) {
+      LoggedHeartbeats logged_heartbeats;
+      bool read_succeeded = this->storage_.ReadTo(logged_heartbeats);
```

Comment:
Added a cooldown period so that GetPayload returns an empty string if it is called within X seconds of a previous nontrivial call (30s by default)
I made this 30s parameter a global and override it in tests